python/feathub/metric_stores/prometheus_metric_store.py

```diff
@@ -0,0 +1,164 @@
+
+DELETE_ON_SHUTDOWN_CONFIG = PROMETHEUS_METRIC_STORE_PREFIX + "delete_on_shutdown"
+DELETE_ON_SHUTDOWN_DOC = """
+Whether to delete metrics from Prometheus when the job finishes. When set to true,
+Feathub will try its best to delete the metrics but this is not guaranteed.
+"""
+
+
+prometheus_metric_store_config_defs: List[ConfigDef] = [
+    ConfigDef(
+        name=HOST_URL_CONFIG,
+        value_type=str,
+        description=HOST_URL_DOC,
+    ),
+    ConfigDef(
+        name=DELETE_ON_SHUTDOWN_CONFIG,
+        value_type=bool,
+        description=DELETE_ON_SHUTDOWN_DOC,
+        default_value=True,
+    ),
+]
+
+
+class PrometheusMetricStoreConfig(MetricStoreConfig):
+    def __init__(self, props: Dict[str, Any]) -> None:
+        super().__init__(props)
+        self.update_config_values(prometheus_metric_store_config_defs)
+
+
+class PrometheusMetricStore(MetricStore):
+    """
+    A metric store reporting metrics to Prometheus through push gateway. Metrics
+    would be reported in the following format:
+
+    - metric name: "feathub_{namespace}_{scope}_{metric_type}". For metrics defined
+                   on Features, scope is "feature".
+    - tags: the metric tags along with the following tags.
+        - feature_name: The name of the host feature
+        - table_name: The name of the sink where the host features would be written to.
+        - job: The namespace of the metric store.
+    """
+
+    def __init__(self, props: Dict) -> None:
+        """
+        :param props: The metric store properties.
+        """
+        super().__init__(props)
+        prometheus_metric_store_config = PrometheusMetricStoreConfig(props)
+        self.host_url = prometheus_metric_store_config.get(HOST_URL_CONFIG)
+        self.delete_on_shutdown = prometheus_metric_store_config.get(
+            DELETE_ON_SHUTDOWN_CONFIG
+        )
+
+    def materialize_metrics_if_any(
+        self,
+        feature_descriptor: TableDescriptor,
+        data_sink: Sink,
+    ) -> Optional[MaterializationDescriptor]:
+        if not any(
+            feature.metrics for feature in feature_descriptor.get_output_features()
+        ):
+            return None
+        return MaterializationDescriptor(
+            feature_descriptor=self._get_metrics_view(feature_descriptor, data_sink),
+            sink=self._get_metrics_sink(data_sink),
+            allow_overwrite=True,
+        )
+
+    def _get_metric_name(self, metric: Metric) -> str:
+        return f"feathub_{self.namespace}_feature_{metric.metric_type}"
+
+    def _get_metric_tags(
+        self, metric: Metric, feature: Feature, data_sink: Sink
+    ) -> Dict[str, str]:
+        return {
+            **metric.get_tags(),
+            "feature_name": feature.name,
+            "table_name": data_sink.name,
+        }
+
+    # TODO: support treating zero window as infinite window in SlidingWindowTransform
+    def _get_metrics_view(
```

Comment:
This method does not seem to depend on any thing specific to Prometheus. Would it be better to move this either to a shared util function or to the parent class `MetricStore`?

Same for other methods that are not specific to Prometheus.

Reply:
The behavior of adding metrics tags as descriptions to features might be exclusive to prometheus store, or at least not suitable for metric stores without support for tags. Thus I would prefer to move common logic to `MetricStore` when we add local or other metric stores.

Reply:
All metric store to be supported by FeatHub should have the `tags` concepts. Otherwise, we would not have the API `Metric#get_tags`. It would be very surprising for a popular metric store not to support `tag` for metrics.

In addition, it seems that the most reasonable way to pass label to the underlying metric store is via the feature's description field, which is exactly what `_get_metrics_view()` does. So it seems the logic in this method is reusable by other metric store.

Reply:
I double-checked Flink's metric reporters and found that they all support tags, so I generally agree with this idea. Some details that differs prometheus/local metric store are as follows.

- Prometheus sink supports using `extraLabels` to configure common labels for all metrics in a feature view. Thus while the default `_get_metric_tags` in `MetricStore` need to specify tags like `table_name` in `_get_metric_tags`, PrometheusMetricStore need not.
- Prometheus uses the label `job` to stands for the `namespace` configuration, while others would still use the label name `namespace` by default.
- LocalMestricStore might use print sink or filesystem sink, neither of which supports parsing Table column comments as labels. Thus this metric store would need to add an extra DerivedFeatureView to merge the comments/descriptions back to metric value before sinking.

I have updated metric_store.py and prometheus_metric_store.py with the above details.

Reply:
Sounds good.